### PR TITLE
refactor(robot-server,notify-server): Fix a minor bug and simplify clients

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

The original implementations of the subscriber and publisher clients managed an async task and send/receive queues. Completely not necessary since zmq is already doing that.

# Changelog

Fix protocol session bug (ThreadManager instead of SynchronousAdapter)
Fix documentation errors.
Remove all the queueing in publisher and subscriber.

# Review requests

does door state still work?

# Risk assessment

Very low.

### CODE REVIEW

**Review by sfoster1 — COMMENTED** [review]
what's up with the `Pipfile.lock` changes?

**Comment by amitlissack** [comment]
> what's up with the `Pipfile.lock` changes?

I modified the Pipfile in notify-server to use editable `shared-data`. Not sure why robot-server got updated. I undid it.

File: notify-server/Pipfile [hunk 1]
```diff
@@ -10,6 +10,7 @@ python_version = "3.7"
 pyzmq = "==19.*,>=19.0.2"
 pydantic = "==1.4"
 opentrons = {editable = true, path = "./../api"}
+opentrons-shared-data = {editable = true, path = "./../shared-data/python"}
```
Comment (SyntaxColoring) [inline notify-server/Pipfile:13]: `pipenv` frightens and perplexes me, so I'm assuming this and `Pipfile.lock` are correct.

File: notify-server/README.rst [hunk 2]
```diff
@@ -67,7 +67,7 @@ To start from the command line:
 
 .. code-block:: bash
 
-   python -m notify_server.subscriber -s tcp://localhost:5555 topic1 topic2
+   python -m notify_server.app_sub -s tcp://localhost:5555 topic1 topic2
```
Comment (SyntaxColoring) [inline notify-server/README.rst:70]: If you're already in the neighborhood fixing readme mistakes, the "Development environment" section incorrectly says that notify-server uses Poetry.

(No need for this to get fixed in this PR, of course.)

File: notify-server/notify_server/clients/publisher.py [hunk 3]
```diff
@@ -29,7 +29,10 @@ async def _send_task(connection: Connection, queue: Queue) -> None:
     try:
         while True:
             entry: QueueEntry = await queue.get()
-            await connection.send_multipart(entry.to_frames())
```
Comment (SyntaxColoring) [inline notify-server/notify_server/clients/publisher.py:32]: This isn't directly related to this PR and you don't have to fix it here, but the name `_send_task()` was confusing to me. I think it means "this is a task whose job it is to send," but it reads as "take a task and send it somewhere." The noun phrase is mistakable as a verb phrase.

File: notify-server/notify_server/clients/publisher.py [hunk 4]
```diff
@@ -29,7 +29,10 @@ async def _send_task(connection: Connection, queue: Queue) -> None:
...
+            try:
+                await connection.send_multipart(entry.to_frames())
+            except Exception:
+                log.exception(f"Failed to send {entry}")
```
Comment (SyntaxColoring) [inline notify-server/notify_server/clients/publisher.py]: Should we really be swallowing all exceptions here? What exceptions do we anticipate from `send_multipart()` and `to_frames()`? Can we be more narrow?

If we have to fail, can we fail somewhere more transparent than this background task? For example, if a user of this API can cause an exception by trying to send a badly-formatted payload, is there a way to raise that error *when they first give us the payload,* instead of here, when we happen to be getting around to sending it?

File: notify-server/notify_server/clients/subscriber.py [hunk 5]
```diff
@@ -11,8 +11,9 @@
 log = logging.getLogger(__name__)
 
 
-def create(host_address: str,
-           topics: typing.Sequence[str]) -> Subscriber:
+async def create(
```
Comment (SyntaxColoring) [inline notify-server/notify_server/clients/subscriber.py]: I wasn't initially sure why this needed to be async. But from our video call, I understand:

* Creating the "background" task [needs](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task:~:text=RuntimeError%20is%20raised%20if%20there%20is%20no%20running%20loop%20in%20current%20thread.) a running event loop.
* Even if it didn't, making this async is a good idea to strongly indicate to the caller that they should be in an async context.

So in principle, changing this to `async` LGTM.

-----

Inside `create_subscriber()`, there's a `sock.connect()` call. If that were a normal network socket, I would be concerned that we're running a potentially long-running blocking operation in this main event loop thread.

I don't know enough about ZMQ sockets to know if that call is actually problematic in our case. What do you think?

File: notify-server/tests/integration/test_pub_sub.py [hunk 6]
```diff
@@ -17,8 +17,10 @@
 async def two_publishers(settings: Settings) -> \
         AsyncGenerator[Tuple[publisher.Publisher, publisher.Publisher], None]:
     """Create two publishers."""
-    pub1 = publisher.create(settings.publisher_address.connection_string())
-    pub2 = publisher.create(settings.publisher_address.connection_string())
+    pub1 = await publisher.create(
+        settings.publisher_address.connection_string())
+    pub2 = await publisher.create(
+        settings.publisher_address.connection_string())
```
Comment (SyntaxColoring) [inline notify-server/tests/integration/test_pub_sub.py]: I really hate our line length limits. :( &lt;/rant&gt;

File: robot-server/robot_server/hardware_wrapper.py [hunk 7]
```diff
@@ -51,6 +53,9 @@ async def init_event_watchers(self):
                       "Cannot initialize robot event watchers.")
             return
 
+        from robot_server.service.dependencies import get_event_publisher
```
Comment (SyntaxColoring) [inline robot-server/robot_server/hardware_wrapper.py]: Is this `import` statement here deliberately, or should it be at the top of the file?

File: robot-server/robot_server/hardware_wrapper.py [hunk 8]
```diff
@@ -21,6 +22,7 @@ class HardwareWrapper:
     def __init__(self):
         self._tm: typing.Optional[ThreadManager] = None
         self._hardware_event_watcher = None
+        self._event_publisher: typing.Optional[Publisher] = None
```
Comment (SyntaxColoring) [inline robot-server/robot_server/hardware_wrapper.py]: I'm a bit unclear on what this whole class is doing, so maybe this is a dumb question, but why are these `Optional`/`None`, instead of being concretely initialized here in `__init__()`?

File: robot-server/robot_server/hardware_wrapper.py [hunk 9]
```diff
@@ -60,20 +65,20 @@ async def init_event_watchers(self):
...
 
-        event_publisher = get_event_publisher()
-        topic = topics.RobotEventTopics.HARDWARE_EVENTS
-        publisher = self._publish_hardware_event.__qualname__
-        event_publisher.send_nowait(topic,
-                                    event.Event(
-                                        createdOn=utc_now(),
-                                        publisher=publisher,
-                                        data=payload))
+        if self._event_publisher:
```
Comment (SyntaxColoring) [inline robot-server/robot_server/hardware_wrapper.py]: See comment above about `self._event_publisher` being optional.

File: robot-server/robot_server/service/dependencies.py [hunk 10]
```diff
@@ -19,16 +19,24 @@
 # The single instance of the SessionManager
 _session_manager_inst = None
 
+# The single instance of Publisher
+_event_publisher_inst = None
```
Comment (SyntaxColoring) [inline robot-server/robot_server/service/dependencies.py]: This wouldn't necessarily be better unless you think it is, but you can avoid a global variable by putting the `_event_publisher_inst` attribute on the `get_event_publisher()` function.

https://stackoverflow.com/a/279597/497934

File: robot-server/robot_server/service/session/session_types/protocol/execution/command_executor.py [hunk 11]
```diff
@@ -89,7 +89,7 @@ def create_worker(configuration: SessionConfiguration,
         protocol_runner = ProtocolRunner(
             protocol=protocol,
             loop=loop,
-            hardware=configuration.hardware.sync,
+            hardware=configuration.hardware,
```
Comment (SyntaxColoring) [inline robot-server/robot_server/service/session/session_types/protocol/execution/command_executor.py:92]: What does this do?

**Review by SyntaxColoring — CHANGES_REQUESTED** [review]
Oops, meant to "request changes" instead of "comment" I think. Computers are hard.

File: robot-server/robot_server/hardware_wrapper.py (hunk 7 above)
Comment (sanni-t) [reply robot-server/robot_server/hardware_wrapper.py]: It creates a circular dependency since the dependencies file imports from hardware wrapper. Hence it's inside here.

File: robot-server/robot_server/hardware_wrapper.py (hunk 8 above)
Comment (sanni-t) [reply robot-server/robot_server/hardware_wrapper.py]: Initialization of the hardware wrapper happens through an async call to `initialize()` during robot-server startup. This initialization fetches the actual hardware control api (or the hw threadmanager) which is required to initialize the hardware event publisher. In case if the hardware api isn't available, ~no publisher will be created~ the simulating hw controller is used and the publisher is hooked to that instead. Hence it's only defining the property here and creating the actual publisher later.

File: notify-server/notify_server/clients/publisher.py (hunk 4 above)
Comment (SyntaxColoring) [reply notify-server/notify_server/clients/publisher.py]: Thinking about this some more: I think exception handling being tough here is a symptom of some funky internal implementation details, which also happen to leak through and cause some funky semantics on `send()`.

This might sound a bit handwavey because I'm rushing out the door. And I'm only like 20% sure how ZMQ works because I haven't read the guide yet. But bear with me.

-----

Our `publisher.Publisher` has an internal `Queue`. This internal queue happens to be of unbounded length (I'm not sure if the unbounded length is intentional, but that's how we make it in `publisher.create()`).

`publisher.Publisher.send()` returns after it places a message on that internal queue. To me, this...seems like an unnecessarily weak primitive? Wouldn't we want it to return when it successfully *gives ZMQ the message*? In a sense, ZMQ is itself a big queue, right? Why, then, are we only using it indirectly, through a feeder queue?

Using an internal feeder queue:

* Hides certain failures until it's too late to handle them properly, as discussed above in my comment about exception handling.
* Circumvents natural backpressure. [Backpressure is a good thing (and hidden buffers can be a bad thing)](https://vorpus.org/blog/some-thoughts-on-asynchronous-api-design-in-a-post-asyncawait-world/#bug-1-backpressure). If, for some reason, ZMQ can't accept messages as quickly as we're submitting them, we should slow down until ZMQ is ready.
* Smells like the kind of thing that will hide deadlocking bugs. I'm vaguely reminded of some Go bugs that I've seen. Handwavey, I know, sorry—but from a high level, you can imagine how a hidden buffer adds nondeterminism to whether a `put()` or `get()` operation will block, and that nondeterminism causes bugs because you have to remember to consider it *every time you `put()` or `get()`*, and that's hard.
* Weakens our message ordering guarantees by a lot. We can no longer say anything about the relative ordering of events submitted by two publishers to a common topic. If we're fine with this, then, okay—but then we should formally document what delivery semantics we *can* rely on, and we'll have to be very careful to build users of our notification system around those guarantees. Because that's the kind of breach of intuition that makes distributed systems hard.

Separately, `publisher.Publisher.send_nowait()` returns immediately, and either submits the message or fails silently depending on whether there's room in the internal queue. Again, I don't think this is a usefully strong primitive. But also, since the queue is unbounded, I think this will never actually fail, and so it's equivalent to `publisher.Publisher.send()`?

-----

tl;dr I think this exception handling question is the code telling us that the internal feeder queue is questionable, and we should listen to the code.

But let me know if this is unclear, and I'll try to be more concrete when I have some more time.

**Review by sanni-t — COMMENTED** [review]
Door state still works 👍

File: notify-server/README.rst (hunk 2 above)
Comment (amitlissack) [reply notify-server/README.rst:70]: Good catch

File: notify-server/notify_server/clients/publisher.py (hunk 4 above)
Comment (amitlissack) [reply notify-server/notify_server/clients/publisher.py]: I made it way more complicated than necessary due to ignorance. The goal was to make sending an event be as non-obtrusive as possible. But I didn't realize that zmq already does that. 

Thanks for thinking about this.

I will remove the queuing.

File: robot-server/robot_server/service/session/session_types/protocol/execution/command_executor.py (hunk 11 above)
Comment (amitlissack) [reply robot-server/robot_server/service/session/session_types/protocol/execution/command_executor.py:92]: `configuration.hardware` is a `ThreadManager`. Its `sync` property returns a `SynchronousAdaptor`. The protocol context requires a synchronous interface to the `API` class.

File: robot-server/robot_server/service/dependencies.py (hunk 10 above)
Comment (amitlissack) [reply robot-server/robot_server/service/dependencies.py]: I like it! Thanks

File: notify-server/notify_server/clients/serdes.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
+"""Implementation of the QueueEntry class."""
```
Comment (SyntaxColoring) [inline notify-server/notify_server/clients/serdes.py]: This docstring looks outdated.

I'd personally just delete this module-level docstring unless you think it has something to add over the method- and class-level docstrings.

File: notify-server/notify_server/clients/publisher.py [hunk 13]
```diff
@@ -34,7 +34,7 @@ async def send(self, topic: str, event: Event) -> None:
 
     def send_nowait(self, topic: str, event: Event) -> Future[Any]:
```
Comment (SyntaxColoring) [inline notify-server/notify_server/clients/publisher.py:35]: Do we need both `send_nowait()` and `send()` still?

File: notify-server/notify_server/network/connection.py [hunk 14]
```diff
@@ -67,15 +68,17 @@ def __init__(self, socket: Socket) -> None:
         """Construct."""
         self._socket = socket
 
-    async def send_multipart(self, frames: List[bytes]) -> None:
+    def send_multipart(self, frames: List[bytes]) -> Future[Any]:
         """Send a multipart message."""
-        await self._socket.send_multipart(frames)
+        # Type ignore is due to zmq not providing type annotation.
+        return self._socket.send_multipart(frames)  # type: ignore
```
Comment (SyntaxColoring) [inline notify-server/notify_server/network/connection.py:74]: Why `def ... -> Future[Any]` and `return`, instead of `async def ... -> None` and `await`?

Not necessarily a negative thing, just wondering.

File: notify-server/notify_server/network/connection.py [hunk 15]
```diff
@@ -67,15 +68,17 @@ def __init__(self, socket: Socket) -> None:
         """Construct."""
         self._socket = socket
 
-    async def send_multipart(self, frames: List[bytes]) -> None:
+    def send_multipart(self, frames: List[bytes]) -> Future[Any]:
```
Comment (SyntaxColoring) [inline notify-server/notify_server/network/connection.py:71]: Why return a `Future[Any]` instead of `await`ing inside the function and returning `None`?

Not necessarily a negative thing, just wondering.

File: notify-server/notify_server/clients/publisher.py (hunk 13 above)
Comment (amitlissack) [reply notify-server/notify_server/clients/publisher.py:35]: We do need both. There is non-async code in our code base that should be able to send notifications.

File: notify-server/notify_server/network/connection.py (hunk 15 above)
Comment (amitlissack) [reply notify-server/notify_server/network/connection.py:71]: For the reason mentioned above. 

The way zmq achieves async is not be using the `async` keyword but by returning `Future` from their method.

File: notify-server/notify_server/clients/serdes.py (hunk 12 above)
Comment (amitlissack) [reply notify-server/notify_server/clients/serdes.py]: Arghhh

File: notify-server/notify_server/clients/publisher.py (hunk 13 above)
Comment (SyntaxColoring) [reply notify-server/notify_server/clients/publisher.py:35]: Makes sense.

Note that this makes the semantics slightly different between async code and non-async code, though. Async code can optionally wait until ["sending is complete"](https://pyzmq.readthedocs.io/en/latest/api/zmq.asyncio.html#socket:~:text=message.-,Returns%20a%20Future%20that%20resolves%20when%20sending%20is%20complete.), while non-async code can't.

This isn't necessarily a problem—just making sure it's intentional.

File: notify-server/notify_server/network/connection.py (hunk 15 above)
Comment (SyntaxColoring) [reply notify-server/notify_server/network/connection.py:71]: Oh interesting. So `send_multipart()` will begin sending the message even before it's `await`ed.

Hmmm. So, `send_multipart()` returns a `Future`, but it seems like `Future`s are [always bound](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.get_loop) to `asyncio`event loops. Which might mean that `send_multipart()` needs to be called with an event loop running, anyway, implying that non-async code can't use it?

File: notify-server/notify_server/clients/publisher.py (hunk 13 above)
Comment (amitlissack) [reply notify-server/notify_server/clients/publisher.py:35]: Yes, of course. You are right. And I believe that it is clear. In the previous iteration the result you waited for was the queueing. It was not possible to wait for the message to be sent.

File: notify-server/notify_server/network/connection.py (hunk 15 above)
Comment (amitlissack) [reply notify-server/notify_server/network/connection.py:71]: `send` and `send_nowait` must be called from a thread with a running loop. In fact, the `Producer` creation must happen in a thread with a running loop. 

That is actually the root issue that led to this PR and its aborted changes. (This is truly a misguided holiday-time PR). The `get_event_publisher` dependency is not marked `async`. FastApi calls non `async` dependencies in a `ThreadPoolExecutor` (in a thread without an event loop).  That is why I made the `create` method `async` back then (remember when I didn't know why when we were chatting?).

The point is that this PR took a different turn and became about simplification of consumers.

Comment (SyntaxColoring) [reply notify-server/notify_server/network/connection.py:71]: Hm, I think I'm lost.

In the above thread, [you said](https://github.com/Opentrons/opentrons/pull/7152#discussion_r550235187) that `send()` and `send_nowait()` were *both* needed because we need to support async callers as well as sync callers.

But here, you're saying "`send` and `send_nowait` must [both] be called from a thread with a running loop"? Doesn't that imply they both only support async callers?

Comment (amitlissack) [reply notify-server/notify_server/network/connection.py:71]: We can talk about this later. I think there's some semantics confusion going on. 

In the mean time. I am closing this PR. It's misguided and dumb.

**Comment by amitlissack** [comment]
Closing this PR. There's nothing good about it.
